org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestManager.xtend

```diff
@@ -38,96 +35,79 @@ import org.eclipse.xtext.util.CancelIndicator
  */
 class RequestManager {
 
-	static val LOGGER = Logger.getLogger(RequestManager)
+	private final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
+	private final Lock r = rwl.readLock();
+	private final Lock w = rwl.writeLock();
 
-	val final MAX_PERMITS = Integer.MAX_VALUE
-	val final WRITE_PERMITS = MAX_PERMITS
-	val final READ_PERMITS = 1
-
-	val sequential = Executors.newSingleThreadExecutor(
-		new ThreadFactoryBuilder().setDaemon(true).setNameFormat("RequestManager-Queue-%d").build
-	)
 	@Inject ExecutorService parallel
 
 	@Inject
 	OperationCanceledManager operationCanceledManager
 
 	val toCancel = new CopyOnWriteArrayList<Cancellable>
 
-	val semaphore = new Semaphore(MAX_PERMITS)
-
 	def void shutdown() {
-		sequential.shutdown()
 		parallel.shutdown()
 		cancel()
 	}
 
-	def <V> V lockRead(()=>V request) {
-		lock(READ_PERMITS, request)
-	}
-
-	def <V> V lockWrite(()=>V request) {
-		lock(WRITE_PERMITS, request)
-	}
-
-	def void cancel() {
-		for (cancellable : toCancel) {
-			cancellable.cancel
-			toCancel -= cancellable
-		}
-	}
-
-	def <V> CompletableFuture<V> runWrite((CancelIndicator)=>V request) {
-		queue(request) [
-			thenApply[lockWrite]
-		]
-	}
-
 	def <V> CompletableFuture<V> runRead((CancelIndicator)=>V request) {
-		queue(request) [
-			thenApplyAsync([lockRead], parallel)
+		val result = new CompletableFuture<V>()
+		writePending.thenRun [			
+			parallel.submit[
+				val cancelIndicator = new RequestCancelIndicator(result)
+				try {
+					r.lock
+					toCancel += cancelIndicator
+					cancelIndicator.checkCanceled
+					result.complete(request.apply(cancelIndicator))
+				} catch (Exception e) {
+					result.completeExceptionally(e)
+				} finally {
+					toCancel -= cancelIndicator
+					r.unlock()
+				}
+			]
 		]
-	}
-
-	protected def <V> CompletableFuture<V> queue(
-		(CancelIndicator)=>V request,
-		(CompletableFuture<()=>V>)=>CompletableFuture<V> requestExecutor
-	) {
-		val origin = new CompletableFuture<RequestCancelIndicator>
-		val requestFuture = requestExecutor.apply(
-			origin.thenApplyAsync([bind(request)], sequential)
-		)
-
-		val cancelIndicator = new RequestCancelIndicator(requestFuture)
-		toCancel += cancelIndicator
-		val result = requestFuture.whenComplete [
-			toCancel -= cancelIndicator
-		]
-
-		origin.complete(cancelIndicator)
 		return result
 	}
-
-	protected def <V> ()=>V bind(RequestCancelIndicator cancelIndicator, (CancelIndicator)=>V request) {
-		cancelIndicator.checkCanceled
-		return [
-			cancelIndicator.checkCanceled
-			request.apply(cancelIndicator)
+	
+	var CompletableFuture<Void> writePending = CompletableFuture.completedFuture(null)
+
+	def <U, V> CompletableFuture<V> runWrite(()=>U nonCancellable, (CancelIndicator, U)=>V request) {
+		val result = new CompletableFuture<V>()
+		writePending = new CompletableFuture<Void>()
+		parallel.submit[
+			try {
+				cancel()
+				w.lock
+				writePending.complete(null)
```

Comment:
This reference to `writePending` could be different from the one created in line 79. In this case there can be readers waiting forever. Make sure that the CompletableFuture that is completed here is really the one that was created for this write request.